reana_server/complexity.py

```diff
@@ -0,0 +1,360 @@
+# -*- coding: utf-8 -*-
+#
+# This file is part of REANA.
+# Copyright (C) 2021 CERN.
+#
+# REANA is free software; you can redistribute it and/or modify it
+# under the terms of the MIT License; see LICENSE file for more details.
+
+"""REANA workflow complexity estimation."""
+
+import os
+from functools import reduce
+
+from reana_commons.job_utils import kubernetes_memory_to_bytes
+from reana_commons.config import MQ_MAX_PRIORITY
+
+from reana_server.status import NodesStatus
+from reana_server.config import REANA_COMPLEXITY_JOBS_MEMORY_LIMIT
+
+
+def get_workflow_min_job_memory(complexity):
+    """Return minimal job memory from workflow complexity.
+
+    :param complexity: workflow complexity list which consists of number of initial jobs and the memory in bytes they require. (e.g. [(8, 1073741824), (5, 2147483648)])
+    :return: minimal job memory (e.g. 1073741824)
+    """
+    if not complexity:
+        return 0
+    return min(complexity, key=lambda x: x[1])[1]
+
+
+def calculate_workflow_priority(complexity):
+    """Calculate workflow priority based on its complexity."""
+    if not complexity:
+        return 0
+    wf_memory = int(reduce(lambda sum, item: sum + item[0] * item[1], complexity, 0))
+    total_cluster_memory = NodesStatus().get_total_memory()
+    if wf_memory > total_cluster_memory:
+        return 1
+    return int(round(1 - wf_memory / total_cluster_memory, 2) * MQ_MAX_PRIORITY)
+
+
+def estimate_complexity(workflow_type, reana_yaml):
+    """Estimate complexity in REANA workflow.
+
+    :param workflow_type: A supported workflow specification type.
+    :param reana_yaml: REANA YAML specification.
+    """
+
+    def build_estimator(workflow_type, reana_yaml):
+        if workflow_type == "serial":
+            return SerialComplexityEstimator(reana_yaml)
+        elif workflow_type == "yadage":
+            return YadageComplexityEstimator(reana_yaml)
+        elif workflow_type == "cwl":
+            return CWLComplexityEstimator(reana_yaml)
+        else:
+            raise Exception(
+                "Workflow type '{0}' is not supported".format(workflow_type)
+            )
+
+    estimator = build_estimator(workflow_type, reana_yaml)
+    try:
+        complexity = estimator.estimate_complexity()
+    except Exception:
+        return []
+    return complexity
+
+
+class ComplexityEstimatorBase:
```

Comment:
I haven't look much into the complexity estimator classes. I assume this was already reviewed as part of the reana-client old PR. Is this right? cc/ @tiborsimko

Reply:
Yes, I tested several scenarios, and the complexity estimator returning information `[(no_jobs_1, memory_1), (no_jobs_2, memory_2), ...] ` for the initial workflow stages (or later stages) was fully OK then :+1: